tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -4135,7 +4135,7 @@ QualType DeclResultIdMapper::getTypeAndCreateCounterForPotentialAliasVar(
   if (const auto *varDecl = dyn_cast<VarDecl>(decl)) {
     // This method is only intended to be used to create SPIR-V variables in the
     // Function or Private storage class.
-    assert(!varDecl->isExternallyVisible() || varDecl->isStaticDataMember());
+    assert(!SpirvEmitter::isExternalVar(varDecl));
```

Comment:
I'm not sure but it seems those asserts are not equivalent?

passes if varDecl->isStaticDataMember is true, same.
But the isExternallyVisible() is not replicated in the new code.

isExternallyVisible returns true if the linkage is >= VisibleNoLinkage while the current checks only allows ExternalLinkage no?
Is that desirable?

Reply:
The old and new assert are not suppose to be equivalent. The existing assert is triggering in cases when it should not be.

The new assert is using the condition in that DXC uses in other places to determine which storage class to place the variable in. This will make all places that check if a variable should be in a storage class other than function or private consistent. That is desirable.

For the other issue: is `isExternalVar` wrong? That function is used to determine if an external variable should be created, so it will never be incorrect for the assert. We also have no reason to believe that a variable is being created in the wrong storage class.

https://github.com/microsoft/DirectXShaderCompiler/blob/2096d1b8054d1de864924ef0c5009f533597b9d4/tools/clang/lib/SPIRV/SpirvEmitter.cpp#L1771-L1773

Reply:
Ok, thanks for the details!